### PR TITLE
feat(ST-99): add endpoint for fetching classroom by id

### PR DESCRIPTION
**Pull request by ghost**
## Description of Change
- Updated the service to fetch classroom based on id and authorized user id.
- Updated the controller for fetching the classroom

## Associated Ticket Link(s)

- https://trello.com/c/4dy7FhPN

## Screenshots / Screen Recordings

N/A

### CODE REVIEW

File: src/classrooms/classrooms.service.ts [hunk 1]
```diff
@@ -65,4 +65,21 @@ export class ClassroomsService {
     // TODO: query for fetching classrooms that a student is enrolled in.
     return classrooms;
   }
+
+  async getClassroomById(id: number, userId: number): Promise<Classroom | null> {
+    const user = await this.em.findOne(User, { id: userId }, { populate: ["teacher"] });
+    if (!user) {
+      throw new NotFoundException("User not found");
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts]: You can use findOneOrFail instead

File: src/classrooms/classrooms.service.ts [hunk 2]
```diff
@@ -65,4 +65,21 @@ export class ClassroomsService {
...
+    }
+    const classroom = await this.em.findOne(Classroom, { id }, { populate: ["teacher"] });
+    if (!classroom) {
+      throw new NotFoundException(`Classroom with ID ${id} not found`);
+    }
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts]: Can use findOneOrFail

File: src/classrooms/classrooms.service.ts [hunk 3]
```diff
@@ -65,4 +65,21 @@ export class ClassroomsService {
...
+    if (user.teacher && classroom.teacher.id === user.teacher.id) {
+      return classroom;
+    }
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts]: If this does match, it should throw forbidden error.

File: src/classrooms/classrooms.service.ts [hunk 4]
```diff
@@ -65,4 +65,17 @@ export class ClassroomsService {
     // TODO: query for fetching classrooms that a student is enrolled in.
     return classrooms;
   }
+
+  async getClassroomById(id: number, userId: number): Promise<Classroom | null> {
+    const user = await this.em.findOneOrFail(User, { id: userId }, { populate: ["teacher"] });
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts]:
```suggestion
    const user = await this.usersRepository.findOneOrFail({ id: userId }, { populate: ["teacher"] });
```

Same comment for other repetitions of this in this file.

File: src/classrooms/classrooms.service.ts [hunk 5]
```diff
@@ -65,4 +65,17 @@ export class ClassroomsService {
...
+
+    const classroom = await this.em.findOneOrFail(Classroom, { id }, { populate: ["teacher"] });
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts]:
```suggestion
    const classroom = await this.classroomsRepository.findOneOrFail({ id }, { populate: ["teacher"] });
```

File: src/classrooms/classrooms.service.ts [hunk 6]
```diff
@@ -65,4 +65,17 @@ export class ClassroomsService {
...
+
+    if (user.teacher && classroom.teacher.id === user.teacher.id) {
+      return classroom;
+    }
+    // TODO: Add check for students enrolled in the classroom
+    // For now, throw ForbiddenException if the user is not the teacher
+    throw new ForbiddenException("You do not have permission to access this classroom");
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts]: The service should always return a classroom and exceptions should be thrown based on some condition.
I think this implementation is flipped.

File: src/classrooms/classrooms.controller.ts [hunk 7]
```diff
@@ -32,4 +32,13 @@ export class ClassroomsController {
...
     return this.classroomsSerializer.serializeMany(classrooms);
   }
+
+  @Get(":id")
+  async getClassroomById(
+    @Param("id") id: number,
+    @CurrentUser() user: { id: number },
+  ): Promise<ClassroomResponseDto> {
+    const classroom = await this.classroomsService.getClassroomById(id, user.id);
+    return this.classroomsSerializer.serialize(classroom);
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.controller.ts:43]: Let's add new line above

File: src/classrooms/classrooms.service.ts [hunk 8]
```diff
@@ -65,4 +67,20 @@ export class ClassroomsService {
...
+    const classroom = await this.classroomsRepository.findOneOrFail(
+      { id },
+      { populate: ["teacher"] },
+    );
+
+    if (user.teacher && classroom.teacher.id !== user.teacher.id) {
+      throw new ForbiddenException("You do not have permission to access this classroom");
+    }
+    // TODO: Add check for students enrolled in the classroom
+    // For now, throw ForbiddenException if the user is not the teacher
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts]: Can remove this comment

File: src/common/entities/users.entity.ts [hunk 9]
```diff
@@ -1,12 +1,22 @@
...
 
 import { EUserType } from "../enums/users.enums";
 import { CustomBaseEntity } from "./custom-base.entity";
 import { Student } from "./students.entity";
 import { Teacher } from "./teachers.entity";
 
-@Entity({ tableName: "users" })
+@Entity({ tableName: "users", repository: () => UserRepository })
 export class User extends CustomBaseEntity {
+  [EntityRepositoryType]?: UserRepository;
```
Comment (ishmam-iqbal-sazim) [inline src/common/entities/users.entity.ts:19]: Let's add new line below

**Review by ishmam-iqbal-sazim — APPROVED** [review]
Left some comments

**Review by ragib-hossain-sazim — APPROVED** [review]

**Pull request merged**